src/java/org/apache/cassandra/repair/RepairTracker.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.repair;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.cassandra.service.StorageServiceMBean;
+
+public class RepairTracker
+{
+    private final ConcurrentHashMap<RepairArguments, RepairRunnable> argsToMostRecentRepair;
```

Comment:
I think it may be worth making this regular hashmaps, and then having all the methods be `synchronized`.  Reason being, I see `argsToMostRecentRepair` as an index, and all changes that happy to `argsToMostRecentRepair` must happen concurrently to both maps. As it stands, I can call `track`, which could put a new value into `argsToMostRecentRepair`, but it doesn't have the lock on `commandToRepairs`, so we could perform a read on `commandToRepairs`, and not see the repair we expect

Reply:
I also think we need to consider how we clean this up afterwards; I'm not sure what `RepairRunnable` holds references to, but this is technically a reference leak.

Reply:
Up to you! How is the command ID generated? Is there any chance that we could re-use an older command id for a newer request?

Reply:
Repair command is a simple counter starting at 1. The only realistic way I think we'd reuse a command ID would be if c* restarted (which would have all this tracking state wiped anyway).

Reply:
Then yeah I think that's fine.

Reply:
Pushed. Though currently we'll only clean up when the tracker gets invoked so it's possible we maintain the reference to 1 RepairRunnable without cleaning it up:
* sidecar triggers repair
* sidecar bounces so it no longer listens for that repairs progress + does not hit our new progress endpoint
* reference will be maintained until c* bounces or sidecar issues another repair, at which point we'll clean up

Do you think it's worth scheduling the cleanup on an executor to keep us 100% clean?

Reply:
discussed offline, having the tracker implement `ProgressListener`